sklearn/linear_model/_base.py

```diff
@@ -287,11 +308,25 @@ def predict(self, X):
 
     def _set_intercept(self, X_offset, y_offset, X_scale):
         """Set the intercept_"""
+
+        xp, _ = get_namespace(X_offset, y_offset, X_scale)
+
         if self.fit_intercept:
             # We always want coef_.dtype=X.dtype. For instance, X.dtype can differ from
             # coef_.dtype if warm_start=True.
-            self.coef_ = np.divide(self.coef_, X_scale, dtype=X_scale.dtype)
-            self.intercept_ = y_offset - np.dot(X_offset, self.coef_.T)
+            coef_ = xp.astype(self.coef_, X_scale.dtype, copy=False)
+            coef_ = self.coef_ = xp.divide(coef_, X_scale)
+
+            if coef_.ndim == 1:
+                intercept_ = y_offset - X_offset @ coef_
+            else:
+                intercept_ = y_offset - X_offset @ coef_.T
+
+            if y_offset.ndim < 1:
+                intercept_ = _item(intercept_, xp=xp)
```

Comment:
I don't fully understand what this does/why it is needed. From the docstring it sounds like `_item` somehow deal with 0 dimensional arrays (scalars, where `x.ndim==0`) and gives me a Python float from such a zero dimensional array.

When I step through it with a debugger then it seems like it gets called with `intercept_` that has `ndim == 0` and no shape. It then reshapes it, and extracts the zero-th element. Which is again a scalar/zero dimensional array. So somehow it feels like we just went in a circle to arrive back where we already were.

Reply:
I just posted a similar feedback.

Furthermore, the `_item` name itself is confusing: in pytorch, the `.item()` method is used to convert a torch scalar tensor into a Python scalar object.